processor/deltatocumulativeprocessor/reorderprocessor/processor.go

```diff
@@ -0,0 +1,198 @@
+)
+
+type Config struct {
+	Hold time.Duration `mapstructure:"hold"`
+	Max  int           `mapstructure:"max_streams"`
+}
+
+func Default() Config {
+	return Config{
+		Hold: 1 * time.Minute,
+		Max:  0, // no limit
+	}
+}
+
+type Processor struct {
+	cfg  Config
+	next consumer.Metrics
+	log  *zap.Logger
+
+	ctx    context.Context
+	cancel func()
+
+	mtx  sync.Mutex
+	meta Meta
+
+	nums Queue[pmetric.NumberDataPoint, pmetric.NumberDataPointSlice]
+	expo Queue[pmetric.ExponentialHistogramDataPoint, pmetric.ExponentialHistogramDataPointSlice]
+	hist Queue[pmetric.HistogramDataPoint, pmetric.HistogramDataPointSlice]
+	summ Queue[pmetric.SummaryDataPoint, pmetric.SummaryDataPointSlice]
+}
+
+func New(cfg *Config, next consumer.Metrics, log *zap.Logger) *Processor {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	proc := Processor{
+		cfg:  *cfg,
+		next: next,
+		log:  log,
+
+		ctx:    ctx,
+		cancel: cancel,
+
+		mtx: sync.Mutex{},
+		meta: Meta{
+			resources: make(map[identity.Resource]pcommon.Resource),
+			scopes:    make(map[identity.Scope]pcommon.InstrumentationScope),
+			metrics:   make(map[identity.Metric]pmetric.Metric),
+		},
+
+		nums: make(Queue[pmetric.NumberDataPoint, pmetric.NumberDataPointSlice]),
+		hist: make(Queue[pmetric.HistogramDataPoint, pmetric.HistogramDataPointSlice]),
+		expo: make(Queue[pmetric.ExponentialHistogramDataPoint, pmetric.ExponentialHistogramDataPointSlice]),
+		summ: make(Queue[pmetric.SummaryDataPoint, pmetric.SummaryDataPointSlice]),
+	}
+
+	return &proc
+}
+
+func (p *Processor) size() int {
+	return len(p.nums) + len(p.expo) + len(p.hist) + len(p.summ)
+}
+
+func (p *Processor) ConsumeMetrics(_ context.Context, md pmetric.Metrics) error {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	var errs error
+	// for rm := range gmetric.ResourceMetrics(md)
+	gmetric.ResourceMetrics(md)(func(rm gmetric.ResourceMetric) bool {
+		id := rm.Ident()
+		// move data to queues
+		switch m := rm.Typed().(type) {
+		case gmetric.Gauge:
+			moveToQueue(p, id, m, p.nums)
+		case gmetric.Sum:
+			moveToQueue(p, id, m, p.nums)
+		case gmetric.Hist:
+			moveToQueue(p, id, m, p.hist)
+		case gmetric.Expo:
+			moveToQueue(p, id, m, p.expo)
+		case gmetric.Summ:
+			moveToQueue(p, id, m, p.summ)
+		}
+
+		// store metadata for later reconstruction
+		p.meta.Store(rm)
+		return true
+	})
+
+	return errs
+}
+
+// moveToQueue moves all datapoints out of the individual metrics and stores
+// them on their respective priority queue.
+//
+// once finished, all metrics have empty datapoint slices, but all metadata is
+// left intact.
+func moveToQueue[M gmetric.Metric[T, S], S gmetric.Slice[S, T], T gmetric.Point[T]](p *Processor, id identity.Metric, m M, q Queue[T, S]) {
+	// for _, dp := range gmetric.All(m.DataPoints())
+	gmetric.All(m.DataPoints())(func(_ int, dp T) bool {
+		id := identity.OfStream(id, dp)
+		if _, has := q[id]; !has && p.size() >= p.cfg.Max {
```

Comment:
Can you add a comment to what this check is doing? We're trying to limit memory usage?